app/src/main/java/com/keylesspalace/tusky/BottomSheetActivity.kt

```diff
@@ -0,0 +1,198 @@
+/* Copyright 2018 Conny Duck
+ *
+ * This file is a part of Tusky.
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>. */
+
+package com.keylesspalace.tusky
+
+import android.content.Intent
+import android.os.Bundle
+import android.support.annotation.VisibleForTesting
+import android.support.design.widget.BottomSheetBehavior
+import android.view.View
+import android.widget.LinearLayout
+import com.keylesspalace.tusky.entity.SearchResults
+import com.keylesspalace.tusky.entity.Status
+import com.keylesspalace.tusky.network.MastodonApi
+import com.keylesspalace.tusky.util.LinkHelper
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.net.URI
+import java.net.URISyntaxException
+
+/** this is the base class for all activities that open links
+ *  links are checked against the api if they are mastodon links so they can be openend in Tusky
+ *  Subclasses must have a bottom sheet with Id item_status_bottom_sheet in their layout hierachy
+ */
+
+abstract class BottomSheetActivity : BaseActivity() {
+
+    lateinit var bottomSheet: BottomSheetBehavior<LinearLayout>
+    var searchUrl: String? = null
+
+    abstract fun getMastodonApi(): MastodonApi
+
+    override fun onPostCreate(savedInstanceState: Bundle?) {
+        super.onPostCreate(savedInstanceState)
+
+        val bottomSheetLayout: LinearLayout = findViewById(R.id.item_status_bottom_sheet)
+            bottomSheet = BottomSheetBehavior.from(bottomSheetLayout)
+            bottomSheet.state = BottomSheetBehavior.STATE_HIDDEN
+            bottomSheet.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
+                override fun onStateChanged(bottomSheet: View, newState: Int) {
+                    when (newState) {
+                        BottomSheetBehavior.STATE_HIDDEN -> cancelActiveSearch()
+                        else -> {
+                        }
+                    }
+                }
+
+                override fun onSlide(bottomSheet: View, slideOffset: Float) {}
+            })
+
+    }
+
+    open fun viewUrl(url: String) {
```

Comment:
Why methods are open? If it's for mocking, we use 'mockito-inline' which doesn't care

Reply:
how can I do this without complicating the code a lot?

Reply:
Wait, I don't understand. You made them open for mocking, right?
You just do `when(activity.methodCall()).doReturn(restult)`. Nothing changes with them being open or not.
I feel like I don't understand something.

Reply:
no, the class is not mocked mith Mockito, but subclassed as FakeBottomSheetActivity

Reply:
Ah, I see now, it's a poor man's Spy. We'll fix it sometime